core_java/grep/src/main/java/ca/jrvs/apps/grep/JavaGrepImp.java

```
@@ -0,0 +1,142 @@
+    if (args.length != 3) {
+      throw new IllegalArgumentException("USAGE: JavaGrep regex rootPath outFile");
+    }
+
+    //Use default logger config
+    BasicConfigurator.configure();
+
+    JavaGrepImp javaGrepImp = new JavaGrepImp();
+    javaGrepImp.setRegex(args[0]);
+    javaGrepImp.setRootPath(args[1]);
+    javaGrepImp.setOutFile(args[2]);
+
+    try {
+      javaGrepImp.process();
+    } catch (Exception ex) {
+      javaGrepImp.logger.error(ex.getMessage(), ex);
+    }
+  }
+
+  @Override
+  public void process() throws IOException {
+    ArrayList<String> matchedLines = new ArrayList<String>();
+    for (File file: listFiles(getRootPath())){
+      for (String line: readLines(file)) {
+        if (containsPattern(line)) {
+          matchedLines.add(line);
+        }
+      }
+    }
+    writeToFile(matchedLines);
+  }
+
+  @Override
+  public List<File> listFiles(String rootDir) {
+    File directory = new File(rootDir);
+    // list of files and directories in directory
+    File[] files = directory.listFiles();
+    List<File> fileList = new ArrayList<File>();
+    for (File file: files) {
+      if (file.isFile()) {
+        fileList.add(file);
+      // recursively add any files in subdirectories of directory
+      } else if (file.isDirectory()) {
+        fileList.addAll(listFiles(file.getAbsolutePath()));
+      }
+    }
+    return fileList;
+  }
+
+  @Override
+  public List<String> readLines(File inputFile) {
+    if (!inputFile.isFile()) {
+      throw new IllegalArgumentException("ERROR: inputFile is not a file!");
+    }
+
+    List<String> lines = new ArrayList<String>();
+    try {
+      BufferedReader bufferedReader = new BufferedReader(new FileReader(inputFile));
+      String currentLine;
+      while ((currentLine = bufferedReader.readLine()) != null) {
+        lines.add(currentLine);
+      }
+      bufferedReader.close();
+    } catch (Exception ex) {
+      logger.error(ex.getMessage(), ex);
```

you might want to throw a runtime exception to stop the program